### PR TITLE
doubly linkedlist

### PR DESCRIPTION
**Pull request by aepaysinger**

### CODE REVIEW

File: doubly_linkedlist.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
+class Dll:
+    def __init__(self, values=None):
+        self.head = None
+        self.previous = None
```
Comment (W-Ely) [inline doubly_linkedlist.py]: The Doubly linked list shouldn't have a previous. It's not needed. You'll need to adjust your code.

File: test_doubly_linkedlist.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    assert dll.head.value == 3
+    assert dll._length == 4
+    assert dll.previous == None
+    assert dll.head.next.value == 8
+
+
+def test_append():
+    dll = Dll([3])
+    dll.append(33)
```
Comment (W-Ely) [inline test_doubly_linkedlist.py]: Can you append to an empty linked list? Prove it with a test,

File: test_doubly_linkedlist.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    assert dll.head.next.next.value == 8
+    assert dll._length == 3
+
+
+def test_pop():
+    dll = Dll([5, 7, 1, 0])
+
+    assert dll.pop() == 0
+    assert dll.head.value == 1
+    assert dll._length == 3
```
Comment (W-Ely) [inline test_doubly_linkedlist.py]: Prove that all the previous are still correct.

File: test_doubly_linkedlist.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+
+def test_remove_tail():
+    dll = Dll([3, 7, 8])
+    dll.remove(3)
+
+    assert dll.head.value == 8
+    assert dll.head.next.value == 7
+    assert dll.head.next.next == None
+    assert dll._length == 2
```
Comment (W-Ely) [inline test_doubly_linkedlist.py]: Both these should prove the .previous is still correct.

File: test_doubly_linkedlist.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
+import pytest
+
+from doubly_linkedlist import Dll
+
+
+def test_dll():
+    dll = Dll([4, 6, 8, 2])
+
+    assert dll.head.value == 2
+    assert dll.head.next.value == 8
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:10]: In this entire file, if it is worth checking what `.next.value` is, it is also worth checking what `.previous.value` is.  Obviously in some cases previous is None rather than having a .value.  But that should only be the head.

File: doubly_linkedlist.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+            for value in values:
+                self.head = Node(value, self.head, None)
+                if self.head.next:
+                    self.head.next.previous = self.head
+                self._length += 1
+
+    def push(self, val):
+        self.head.next = self.head
+        self.head = Node(val, self.head, None)
+        self._length += 1
```
Comment (W-Ely) [inline doubly_linkedlist.py]: To me this doesn't look like it's adjusting the .previous's correctly, and your tests are insufficient to convince me.

File: doubly_linkedlist.py [hunk 7]
```diff
@@ -2,20 +2,23 @@ class Dll:
...
         self.head = None
         self._length = 0
+        self.tail = None
         if values:
             for value in values:
                 self.head = Node(value, self.head, None)
                 if self.head.next:
                     self.head.next.previous = self.head
+                if self.head.next == None:
+                    self.tail = self.head
```
Comment (W-Ely) [inline doubly_linkedlist.py]: else?

File: doubly_linkedlist.py [hunk 8]
```diff
@@ -26,14 +29,17 @@ def append(self, val):
                 if current.next == None:
                     if current == self.head:
                         self.head.next = Node(val, None, self.head)
+                        self.tail = self.head.next
```
Comment (W-Ely) [inline doubly_linkedlist.py]: The whole point of having a tail is to avoid needing to loop to get to the tail end.

File: doubly_linkedlist.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+class Node:
+    def __init__(self, value, next, previous):
+        self.value = value
+        self.next = next
+        self.previous = previous
+
+
+dll = Dll([2, 3, 4])
+print(dll.__init__)
```
Comment (W-Ely) [inline doubly_linkedlist.py]: if __name__ == "__main__": or remove.

File: test_doubly_linkedlist.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+def test_push():
+    dll = Dll([5, 7, 8])
+    dll.push(3)
+
+    assert dll.head.value == 3
+    assert dll._length == 4
+    assert dll.head.previous == None
+    assert dll.head.next.value == 8
+    assert dll.head.next.previous.value == 3
+    assert dll.tail.value == 5
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:25]: tail.previous == None

File: test_doubly_linkedlist.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def test_push_empty():
+    dll = Dll()
+    dll.push("Head")
+
+    assert dll.head.value == "Head"
+    assert dll.tail.value == "Head"
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:33]: .previous

File: test_doubly_linkedlist.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def test_append():
+    dll = Dll([])
+    dll.append(33)
+
+    assert dll.head.value == 33
+    assert dll.head.next == None
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:43]: .previous

File: test_doubly_linkedlist.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    assert dll.tail.value == 33
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:44]: .previous

File: test_doubly_linkedlist.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+    assert dll._length == 1
+
+    dll.append(8)
+
+    assert dll.head.value == 33
+    assert dll.head.next.value == 8
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:52]: .previous

File: test_doubly_linkedlist.py [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+    assert dll._length == 2
+    assert dll.tail.value == 8
+    assert dll.tail.previous.value == 33
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:55]: .next

File: test_doubly_linkedlist.py [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def test_pop():
+    dll = Dll([5, 7, 1, 0])
+
+    assert dll.pop() == 0
+    assert dll.head.value == 1
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:71]: .previous

File: test_doubly_linkedlist.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+    assert dll._length == 3
+    assert dll.head.next.next.previous.value == 7
+    assert dll.head.previous == None
+    assert dll.tail.value == 5
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:76]: tail.next

Comment (W-Ely) [reply test_doubly_linkedlist.py:76]: tail.previous

File: test_doubly_linkedlist.py [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+
+def test_shift():
+    dll = Dll([4, 8, 1])
+
+    assert dll.shift() == 4
+    assert dll._length == 2
+    assert dll.head.next.value == 8
+    assert dll.head.next.previous.value == 1
+    assert dll.tail.value == 8
+    assert dll.tail.next == None
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:111]: .previous

File: test_doubly_linkedlist.py [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+        dll.pop()
+    assert exc_info.value.args[0] == "Empty Dll"
+
+
+def test_shift():
+    dll = Dll([4, 8, 1])
+
+    assert dll.shift() == 4
+    assert dll._length == 2
+    assert dll.head.next.value == 8
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:108]: head.previous

File: test_doubly_linkedlist.py [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+    assert dll.head.next.previous.value == 1
+    assert dll.tail.value == 8
+    assert dll.tail.next == None
+
+
+def test_shift_empty():
+    dll = Dll([])
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:118]:
```suggestion
    dll = Dll()
```

File: test_doubly_linkedlist.py [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+    assert exc_info.value.args[0] == "Empty Dll, no items to shift"
+
+
+def test_remove_head():
+    dll = Dll([3, 7, 5])
+    dll.remove(5)
+
+    assert dll.head.value == 7
+    assert dll._length == 2
+    assert dll.tail.value == 3
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:134]: .previous(s)

**Review by W-Ely — COMMENTED** [review]
For every .next in your tests I expect to see a .previous also.  I'm going to count.

File: test_doubly_linkedlist.py (hunk 10 above)
Comment (aepaysinger) [reply test_doubly_linkedlist.py:25]: wouldn't tail.previous == the previous node and then tail.next == None?

File: test_doubly_linkedlist.py (hunk 15 above)
Comment (aepaysinger) [reply test_doubly_linkedlist.py:55]: isnt that covered in line 48? dll.head.next.value

**Comment by aepaysinger** [comment]
most of these tests past w.o changing but one of them did not so I see the value in testing all. how do you know it works if you don't test it?

File: test_doubly_linkedlist.py (hunk 10 above)
Comment (W-Ely) [reply test_doubly_linkedlist.py:25]: Yup. Include it in the tests.

**Comment by W-Ely** [comment]
> how do you know it works if you don't test it?

Exactly.  Also as code becomes more complicated you'll find that it is also very important when reviewing the code to have good test coverage.  For example, if you have a really complicated method that is supposed to do a thing, I can spend an hour trying to fully understand the code and be confident it does the thing, or I can read the test that shows empirically that it does the thing and move on.  This does mean that the tests have to also be thorough enough to instill the same confidence as the code itself would.

^ Coverage above means more than running the lines of code.  Running the lines isn't the same as testing the lines.

**Comment by W-Ely** [comment]
Revisit the 'append' method, now that you have a tail, you shouldn't have to iterate from the head.

File: test_doubly_linkedlist.py [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+    dll = Dll()
+    dll.push("Head")
+
+    assert dll.head.value == "Head"
+    assert dll.tail.value == "Head"
+    assert dll.tail.previous == None
+    assert dll.head.previous == None
+
+
+def test_append():
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:38]: Append one more number here, the 2 you have isn't catching the issues in your code.

File: test_doubly_linkedlist.py [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+
+    dll.append(4)
+
+    assert dll.head.value == 33
+    assert dll.tail.value == 4
+    assert dll.head.next.value == 8
+    assert dll.tail.previous.value == 8
+
+
+def test_pop():
```
Comment (W-Ely) [inline test_doubly_linkedlist.py:67]: This test should pop everything off, going from some to none is the edge case your pop method doesn't account for.

**Review by W-Ely — APPROVED** [review]
